src/lexer.coffee

```diff
@@ -594,27 +537,33 @@ exports.Lexer = class Lexer
           first_column: locationToken[2].first_column
           last_line: locationToken[2].first_line
           last_column: locationToken[2].first_column
-      if tag is 'TOKENS'
-        # Push all the tokens in the fake 'TOKENS' token.  These already have
-        # sane location data.
-        @tokens.push value...
-      else if tag is 'NEOSTRING'
-        # Convert NEOSTRING into STRING
-        token[0] = 'STRING'
-        token[1] = @makeString value, '"', heredoc
-        @tokens.push token
-      else
-        @error "Unexpected #{tag}"
+      switch tag
+        when 'TOKENS'
+          # Push all the tokens in the fake 'TOKENS' token. These already have
+          # sane location data.
+          @tokens.push value...
+        when 'NEOSTRING'
+          # Convert 'NEOSTRING' into 'STRING'.
+          converted = fn token[1], i
+          # Optimize out empty strings. We ensure that the tokens stream always
+          # starts with a string token, though, to make sure that the result
+          # really is a string. TODO: Change this according to #3529.
```

Comment:
I don't like leaving TODOs around. This isn't a particularly important one either, so let's remove it.